src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+      if ((int)i == index) {
+        ret.push_back(DoAdd(origin->fields[i], increment));
+      } else {
+        ret.push_back(origin->fields[i]);
+      }
+    }
+    return Tuple(ret);
+  }
+
+  void BindAndEmit(Var v, Expr e) {
+    if (adjoint_expr_to_var_.count(e)) {
+      e = adjoint_expr_to_var_[e];
+    } else {
+      adjoint_expr_to_var_.Set(e, v);
+      // e->checked_type_ = v->checked_type();
+      // e->shape_ = v->shape();
+    }
+    if (v->IsInstance<DataflowVarNode>()) {
+      builder_->Emit(VarBinding(v, e));
+    } else {
+      builder_->EmitOutput(VarBinding(v, e));
+    }
+  }
+
+  void CheckTarget(const Expr& e) {
+    ICHECK(!e->IsInstance<DataflowVarNode>()) << "not an output node";
+    ICHECK(e->checked_type().as<DynTensorTypeNode>()) << "target must be a DynTensorType";
+    ICHECK(e->shape().as<ShapeExprNode>()) << "error when getting target shape";
+    const auto* shape_node = e->shape().as<ShapeExprNode>();
+    ICHECK(shape_node->values.size() == 0) << "target must be a scalar";
+  }
+
+  void InitGrad(const Var& adjoint_var, const Var& var) {
+    static const Op& ones_op = Op::Get("relax.ones");
+
+    ObjectPtr<InitAttrs> attrs = make_object<InitAttrs>();
+    auto type = Downcast<DynTensorType>(var->checked_type());
+    attrs->dtype = type->dtype;
+
+    const Expr& init = Call(ones_op, {var->shape()}, Attrs(attrs));
+    adjoint_expr_map_.Set(var, init);
+  }
+
+  // the differentiation target
+  Var target_var_;
+  // the arguments to differentiate
+  Array<Var> require_grads_;
+
+  // var to its adjoints var
+  Map<Var, Var> adjoint_var_map_;
+  // var to its adjoint expr
+  Map<Var, Expr> adjoint_expr_map_;
+  // trace binding
+  Map<Expr, Var> adjoint_expr_to_var_;
+  // track zeros introduced
+  std::unordered_set<Expr, ObjectPtrHash, ObjectPtrEqual> zeros_tracker_;
+
+  // gop map
+  const OpAttrMap<FPrimalGradient> gradient_op_map_ =
+      Op::GetAttrMap<FPrimalGradient>("FPrimalGradient");
+};
+
+/*!
+ * \brief A simple reverse-mode auto differentiation.
+ * \param m The module
+ * \param var The GlobalVar of the specific function
+ * \param require_grad_names The relax variables which need adjoints. Must be inputs.
+ * \return The module after AD.
+ */
+IRModule Gradient(IRModule m, const GlobalVar& var, const Array<Var>& require_grads) {
+  BaseFunc base_func = m->Lookup(var);
+  if (auto* n = base_func.as<FunctionNode>()) {
+    auto f_before = GetRef<Function>(n);
+    Array<Var> require_grads_var;
+    for (auto input : require_grads) {
+      ICHECK(std::find(n->params.begin(), n->params.end(), input) != n->params.end())
+          << "function " << var->name_hint << " has no var named " << input->name_hint();
+      require_grads_var.push_back(input);
+    }
+
+    IRModuleNode* new_module_node = m.CopyOnWrite();
+    auto new_module = GetRef<IRModule>(new_module_node);
+    auto mutator = GradientMutator(new_module, require_grads_var);
```

Comment:
Seems to me that just passing the original IRModule to the mutator constructor is fine. I don’t think we need use CopyOnWrite here?

Reply:
It refers to `src/relax/transform/bind_params.cc`